cpp/arcticdb/storage/s3/nfs_backed_storage-inl.hpp

```diff
@@ -133,17 +133,16 @@ inline void NfsBackedStorage::do_update(Composite<KeySegmentPair>&& kvs, UpdateO
     s3::detail::do_update_impl(std::move(enc), root_folder_, bucket_name_, s3_client_, NfsBucketizer{});
 }
 
-template<class Visitor>
-void NfsBackedStorage::do_read(Composite<VariantKey>&& ks, Visitor&& visitor, ReadKeyOpts opts) {
-    auto func = [v = std::forward<Visitor>(visitor)] (const VariantKey& k, Segment&& seg) mutable {
-        v(unencode_object_id(k), std::move(seg));
+inline void NfsBackedStorage::do_read(Composite<VariantKey>&& ks, const ReadVisitor& visitor, ReadKeyOpts opts) {
+    auto func = [visitor] (const VariantKey& k, Segment&& seg) mutable {
```

Comment:
How safe is this? Up 'till now the `func` lambda and the `do_read` function were owning the visitor. Now the visitor is captured as non-owning reference. It's quite easy to make a mistake somewhere along the chain so that the Visitor reference gets deleted while the lambda is still holding reference to it.

Reply:
For now, there are just two `ReadVisitor` defined respectively in [`Storage::read`](https://github.com/man-group/ArcticDB/pull/670/files#diff-e9592f894f71949908c88db11096271fa492cdf2dda147ae4b2daf01cac640c1R120-R123) and in [`Library::read`](https://github.com/man-group/ArcticDB/pull/670/files#diff-6dfbf8b64a20f983135e603f91383f5a63dbe766123124518db124d915d5a2e0R117-R119) which are simply passed down.

So we do not face this problem, but it might be possible to have a deletion. Although I think it is hardly probable, I would very welcome any improvement preventing this issue.

Reply:
I think this might actually improve things in the future, because sometimes we have a problem when for example we want to do storage fall-through (which we don't do much at the moment, but could be very useful in the future), in that we want to try on one storage and if we fail try on another - this is complicated if we've moved the visitor as we have to clone it first, imposing a cost even in the case where the first storage read succeeds.